feat(vis-type: scatter): improve rendering of scatter plot

dvmoritzschoefl opened the pull request:
Closes #522  ,https://github.com/datavisyn/reprovisyn/issues/1870, https://github.com/datavisyn/communication_bi_ordino/issues/76, https://github.com/datavisyn/communication_bi_ordino/issues/77,
https://github.com/datavisyn/reprovisyn/issues/2028,
https://github.com/datavisyn/communication_bi_ordino/issues/79

### Developer Checklist (Definition of Done)

**Issue**

- [ ] All acceptance criteria from the issue are met
- [ ] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [ ] Requires UI/UX/Vis review
  - [ ] Reviewer(s) are notified (_tag assignees_)
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [ ] Branch is up-to-date with the branch to be merged with, i.e., develop
- [ ] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [ ] Descriptive title for this pull request is provided (will be used for release notes later)
- [ ] Reviewer and assignees are defined
- [ ] Add type label (e.g., *bug*, *feature*) to this pull request
- [ ] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [ ] The PR is connected to the corresponding issue (via `Closes #...`)
- [ ] [Summary of changes](#summary-of-changes) is written


### Summary of changes

- fix regression stats

### Screenshots


### Additional notes for the reviewer(s)

-  
Thanks for creating this pull request 🤗

oltionchampari commented:
With commit https://github.com/datavisyn/visyn_core/pull/530/commits/6414a56fc3de8fa59e8dbb7cb90ff13cdb24a60a

I added the initial height of the scatter to be always 100%

The issue is that the in ordino facets are filtered out from the current facet column but the height of the plots does not update to fill the available space
This happens only when you run the production build.

thinkh reviewed (changes requested):
@dvmoritzschoefl I've tested this in Ordino and it works very good! However, the plot doesn't seem to rescale when opening the configuration sidepanel (compared to the other visualization types) and is also not re-rendered reliably when the view/container size changes. The view resize could be also a problem in Ordino where @oltionchampari could assist you.

[ordino-scatterplot-resize.webm](https://github.com/user-attachments/assets/4108d684-c5e7-416f-8945-62f2372f9b3f)

dvmartinweigl commented:
- [ ] when changing the number of numerical columns the plot does not scale horizontally or vertically
![image](https://github.com/user-attachments/assets/ffa9abdb-a9f8-45fc-9ed4-da237843f5a2)

- [ ] we had the distribution along the diagonal when plotting multiple numerical values
before:
![image](https://github.com/user-attachments/assets/84636e71-16b8-47df-9284-2a9d0a657710)
after:
![image](https://github.com/user-attachments/assets/ce0b74ae-7250-427b-b31f-63ee83797c9c)

- [ ] better tooltip with x and y axis labels
![image](https://github.com/user-attachments/assets/8793ead0-5d47-4216-ad82-a5b7e2a5c0c6)

- [ ] stats hover event does not work anymore

- [ ] disable the facets when multiple numerical columns selected. Currently nothing happens when selecting facets

dvchristianbors commented:
After discussing, we agreed that the splom's diagonal histogram will not be done at this point. I created a [follow-up ticket](https://github.com/datavisyn/visyn_core/issues/552) for it to add it back in in the future.

thinkh commented:
I merged the resize fix into this branch and it works now as expected. Thanks for the fix @dvmoritzschoefl.

While testing the updates in Ordino, I noticed two more things:

1. Selected points are not orange anymore
2. Axis titles are missing

**Before the rewrite**

![image](https://github.com/user-attachments/assets/a4ec5ae0-7bab-4feb-9c28-9744751f3e33)


**This PR**

![image](https://github.com/user-attachments/assets/0be119cd-25e0-40da-b8c8-179161f92aad)


Can you please have a look at it on Monday? 


> After discussing, we agreed that the splom's diagonal histogram will not be done at this point. I created a https://github.com/datavisyn/visyn_core/issues/552 for it to add it back in in the future.

@dvchristianbors What did you agree about the remaining points from @dvmartinweigl feedback. Will @dvmoritzschoefl address the feedback in this PR?

dvchristianbors commented:
I just tested creating a screenshot and I get the following eror:

```
Could not find plot div to capture screenshot
console.error @ app.8799d0e7.js:8
```

Probably some prop is not passed to the header (according to @oltionchampari ).

dvmoritzschoefl commented:
Fixed and verified a few things:
- Download is working for me
- Changed point color to primary color
- Changed selection behavior to retain color
- Enabled stats callback again
- Added title axes

The tooltip does not work for splom traces (since its the same point for plotly). I made it uniform with the other traces thats why x/y are displayed without the description. Should this be enabled for the solo trace @dvchristianbors ?

dvmoritzschoefl commented:
> @dvmoritzschoefl Thanks for integrating the latest feedback. It looks now very good.
> 
> I only found one point where I'm unsure whether or not it is intentional or might be an oversight.
> 
> When I select a facet column, and zoom with the mouse wheel in one scatter plot the remaining plots do not update. BUT once I select something all other plots are updating.
> 
>  facet-scatterplot-update-after-selection.webm 
> In comparison, all axes are updated in the scatter plot matrix (SPLOM) instantly when zooming in or out.
> 
>  scatterplot-matrix-instant-update.webm 
> Would it be easy to implement an instant (or debounced) update of the other plots in the faceted mode? Or if not could we disable the update of the remaining plots after selecting some points?

This is intentional, since the plotly implementation for syncing axis is awfully slow, and therefore even with less than 1000 points it begins to get laggy. (You can enable syncing axis by commenting in the part where i mentioned this exact issue in the layout generation (ScatterVis)). If this feature is required I can reenable it but I think @dvchristianbors said that he preferred a smooth experience over it.

thinkh commented:
Thanks for the quick reply and the explanation. I've tested this with [100.000 points](https://github.com/datavisyn/visyn_core/blob/84a3184ee36ce8540f097a5985f0d9f610051b39/src/vis/stories/Vis/Scatter/ScatterRandom.stories.tsx#L98) in storybook and it feels like that the performance is fine when interacting with 10 facets.

[scatterplot-facets-sync-axes.webm](https://github.com/user-attachments/assets/8b756cb7-0ac3-4dcd-89ff-fdf8936b0489)

I noticed that that the performance is worse when rendering the 100k points in the single plot (see beginning of the video) and observed that the performance seems to depend on the rendered canvas size. When I interact with the same example but with a maximized browser window the performance gets noticeably worse when rendering all points. Once I zoom in, and only a subset is rendered, it is fast again.

[scatterplot-facets-sync-axes-large.webm](https://github.com/user-attachments/assets/6d9b1feb-8c5a-4991-8bf8-5a09ce664713)

Interestingly, the performance does not increase for me when disabling the synced axes again (= your PR). The facet I'm interacting with is still very laggy and feels even slower than in the synced-axes example.

[scatterplot-facets-unsynced-axes-large.webm](https://github.com/user-attachments/assets/b742786e-4ea7-43a2-a7ed-1ef692e11610)


Given these results, I'd be in favor of enabling the synced axes for facets.

@dvmoritzschoefl and @dvchristianbors What do you think about enabling the synced axes after seeing the results (and maybe testing it again yourself)?

dvmoritzschoefl commented:
> Thanks for the quick reply and the explanation. I've tested this with [100.000 points](https://github.com/datavisyn/visyn_core/blob/84a3184ee36ce8540f097a5985f0d9f610051b39/src/vis/stories/Vis/Scatter/ScatterRandom.stories.tsx#L98) in storybook and it feels like that the performance is fine when interacting with 10 facets.
> 
>  scatterplot-facets-sync-axes.webm 
> I noticed that that the performance is worse when rendering the 100k points in the single plot (see beginning of the video) and observed that the performance seems to depend on the rendered canvas size. When I interact with the same example but with a maximized browser window the performance gets noticeably worse when rendering all points. Once I zoom in, and only a subset is rendered, it is fast again.
> 
>  scatterplot-facets-sync-axes-large.webm 
> Interestingly, the performance does not increase for me when disabling the synced axes again (= your PR). The facet I'm interacting with is still very laggy and feels even slower than in the synced-axes example.
> 
>  scatterplot-facets-unsynced-axes-large.webm 
> Given these results, I'd be in favor of enabling the synced axes for facets to be consistent with the SPLOM and improve the usability.
> 
> @dvmoritzschoefl and @dvchristianbors What do you think about enabling the synced axes after seeing the results (and maybe testing it again yourself)?

Just to be sure, do you have a dedicated GPU in your laptop and **not** battery saving enabled? This has the biggest impact on performance in gl traces.

As for the performance getting better when zooming in or having a smaller canvas size. This is what I would expect. In PSE the limiting factor always was the fragment shader (which gets called for each pixel on each point), so less points = better performance and more canvas size = worse performance.

I tested the storybook like you did, and it seems the shared axis work fine now (I really do not know why they were so laggy during development, maybe it had to do with the developer console updating all the axis). However I could not reproduce the issue with the single plot. What I observed is (with a dedicated gpu and performance mode):

- Single plot and facet mode works smooth with 100k points and have the same performance (which is what I would expect since we render the same amount of points)
- SPLOM has the worst performance, since we render 9 times the number of points
- Axis sharing does not have a performance impact in facet mode

So I will enable axis sharing again?

dvchristianbors commented:
> > Thanks for the quick reply and the explanation. I've tested this with [100.000 points](https://github.com/datavisyn/visyn_core/blob/84a3184ee36ce8540f097a5985f0d9f610051b39/src/vis/stories/Vis/Scatter/ScatterRandom.stories.tsx#L98) in storybook and it feels like that the performance is fine when interacting with 10 facets.
> > scatterplot-facets-sync-axes.webm
> > I noticed that that the performance is worse when rendering the 100k points in the single plot (see beginning of the video) and observed that the performance seems to depend on the rendered canvas size. When I interact with the same example but with a maximized browser window the performance gets noticeably worse when rendering all points. Once I zoom in, and only a subset is rendered, it is fast again.
> > scatterplot-facets-sync-axes-large.webm
> > Interestingly, the performance does not increase for me when disabling the synced axes again (= your PR). The facet I'm interacting with is still very laggy and feels even slower than in the synced-axes example.
> > scatterplot-facets-unsynced-axes-large.webm
> > Given these results, I'd be in favor of enabling the synced axes for facets to be consistent with the SPLOM and improve the usability.
> > @dvmoritzschoefl and @dvchristianbors What do you think about enabling the synced axes after seeing the results (and maybe testing it again yourself)?
> 
> Just to be sure, do you have a dedicated GPU in your laptop and **not** battery saving enabled? This has the biggest impact on performance in gl traces.
> 
> As for the performance getting better when zooming in or having a smaller canvas size. This is what I would expect. In PSE the limiting factor always was the fragment shader (which gets called for each pixel on each point), so less points = better performance and more canvas size = worse performance.
> 
> I tested the storybook like you did, and it seems the shared axis work fine now (I really do not know why they were so laggy during development, maybe it had to do with the developer console updating all the axis). However I could not reproduce the issue with the single plot. What I observed is (with a dedicated gpu and performance mode):
> 
> * Single plot and facet mode works smooth with 100k points and have the same performance (which is what I would expect since we render the same amount of points)
> * SPLOM has the worst performance, since we render 9 times the number of points
> * Axis sharing does not have a performance impact in facet mode
> 
> So I will enable axis sharing again?

I tested the performance in Xaira's deployment, as you described, performance of the synced axes when faceting is fine, when rendering 40k points. I would suggest enabling axis sharing.

puehringer commented on package.json:
```diff
@@ -106,6 +106,8 @@
     "d3-force-boundary": "^0.0.3",
     "d3-hexbin": "^0.2.2",
     "d3v7": "npm:d3@^7.9.0",
+    "echarts": "^5.5.1",
+    "echarts-gl": "^2.0.9",
```
Remove

thinkh commented:
Current status for the SPLOM:

![image](https://github.com/user-attachments/assets/602a9b0a-33f7-4ddd-be01-6ae002652f44)

dvmartinweigl commented:
With the lower triangle in the splom, the regression lines seem to be broken
![image](https://github.com/user-attachments/assets/0c6ea711-7c03-4e59-87b8-63698fa64739)

dvmoritzschoefl commented:
> With the lower triangle in the splom, the regression lines seem to be broken ![image](https://private-user-images.githubusercontent.com/101096478/375030441-0c6ea711-7c03-4e59-87b8-63698fa64739.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NDA1NTMsIm5iZiI6MTcyODU0MDI1MywicGF0aCI6Ii8xMDEwOTY0NzgvMzc1MDMwNDQxLTBjNmVhNzExLTdjMDMtNGU1OS04N2I4LTYzNjk4ZmE2NDczOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxMFQwNjA0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYmI4YWE1ZTVlZmU5MGFiMTUzYWUwMmM2ZDg5M2Q1MjE4ZWNmZWZkMDcyYjI1ZGU2NzUyN2FlZjc1YmVlZWY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.E-RPrPVYOsjCyoB_grT4qVEVoMNXY_Jc20H8yAw2yes)

thanks for mentioning, it still calculated regression lines for the whole matrix. this is fixed now

thinkh commented:
@dvmoritzschoefl as discussed in call:

1. Distribute subplots equally for the available width
![image](https://github.com/user-attachments/assets/2fde0849-ab6a-41de-8565-e5f95d267542)

2. Changing the drag mode in the vis config is not reflected in the plot

dvmartinweigl commented:
Could we maybe expose a prop in the scattervis config to change the type of splom.
What do you think @dvmoritzschoefl 

Something like this:
`splomMode`:
- `lowerTriangle`
- `fullMatrix`
- `hideDiagonal`
- `histogram` (if added in future)

dvmoritzschoefl commented:
> Could we maybe expose a prop in the scattervis config to change the type of splom. What do you think @dvmoritzschoefl
> 
> Something like this: `splomMode`:
> 
> * `lowerTriangle`
> * `fullMatrix`
> * `hideDiagonal`
> * `histogram` (if added in future)

I would not make it more complicated than needed for now. I think if we cannot have histograms, then the lower triangle mode is the best bet (and also the best performing one). I would add it if someone needs this at some point.

dvmoritzschoefl commented:
> @dvmoritzschoefl as discussed in call:
> 
> 1. Distribute subplots equally for the available width
>    ![image](https://private-user-images.githubusercontent.com/5851088/375391638-2fde0849-ab6a-41de-8565-e5f95d267542.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg1NzM2NzEsIm5iZiI6MTcyODU3MzM3MSwicGF0aCI6Ii81ODUxMDg4LzM3NTM5MTYzOC0yZmRlMDg0OS1hYjZhLTQxZGUtODU2NS1lNWY5NWQyNjc1NDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTBUMTUxNjExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjliYjU4NTYxYTU4MGExOGM0ODM5ZjhmNjI0ZDAwN2MyYWNiMjhiMjY2ZWYyNTRmZThjY2U0MmU4NWZmMmQyNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vb7ayMw_q7F_zSiCFSqY3Mwgt4aPDxuheGBgJAk75hs)
> 2. Changing the drag mode in the vis config is not reflected in the plot

The number of columns and rows is now calculated by the following rule:
- per 400 pixels one more columns comes (until a maximum of 4 columns) -> if less facets than columns are present then the number of columns is the same as the number of facets
- the number of rows is facetnumber / columnnumber
while not perfect for sure this works well for the 1-10 facet cases

as for the plot not updating and the dragmode having no effect. i think this is an ordino problem (as discussed in call). i tried it in the storybook and when using the config in an external state and updating there it works as expected

dvchristianbors commented:
One more small remark is regarding the legend and coloring of unknown categorical values

![image](https://github.com/user-attachments/assets/669f3c97-c3ea-4635-bbc2-192657ad26ff)

The categories are not 'sorted' (previously, the 'Unknown' category always came last), and the vis_neutral_color is not used, as recommended:

```ts
/**
 * Neutral color (e.g., histogram in scatterplot matrix and should be also used for "Unknown" categorical values)
 */
export const VIS_NEUTRAL_COLOR = '#71787E';
```
https://github.com/datavisyn/visyn_core/blob/cb/scatter-bar-plot-improvements/src/vis/general/constants.ts#L31

dvmartinweigl commented:
Something seems wrong with the stats calculations. Adding 4 numeric columns (all gene expression) to the config results in the last row of the lower triangle having NaN values for pValue, r² ....
Could it be that there is an error in the data getting passed to `fitRegressionLine` @dvmoritzschoefl ?

![image](https://github.com/user-attachments/assets/00086a18-a2f1-4e55-bd14-e9dd97d74ccc)

thinkh reviewed (dismissed):
Thanks for the performance improvements and addressing all the various feedback. We have tested it in multiple apps and it is ready to merge.

thinkh commented:
@dvmoritzschoefl I saw that the Playwright tests are failing for this branch. in comparison the develop branch succeeded with the latest merge. Please check the tests and fix them. Thanks.

thinkh reviewed (approved):
I fixed the failing Playwright tests by adding and using the correct selectors for the legend.

thinkh commented:
The download Playwright tests are currently timing-out but all other tests are working as expected.
